internal/jimm/controller.go

```diff
@@ -393,133 +393,231 @@ func (j *JIMM) GetUserControllerAccess(ctx context.Context, user *openfga.User,
-	model.CloudCredential = cloudCredential
+	m.model.CloudCredentialID = cloudCredential.ID
+	m.model.CloudCredential = cloudCredential
+
+	return nil
+}
 
+func (m *modelImporter) setModelCloud(ctx context.Context) error {
 	// fetch the cloud used by the model
-	cloud := dbmodel.Cloud{
-		Name: cloudCredential.CloudName,
+	cloudTag, err := names.ParseCloudTag(m.modelInfo.CloudTag)
+	if err != nil {
+		return err
 	}
-	err = j.Database.GetCloud(ctx, &cloud)
+	cloud := dbmodel.Cloud{Name: cloudTag.Id()}
+	err = m.jimm.Database.GetCloud(ctx, &cloud)
 	if err != nil {
 		zapctx.Error(ctx, "failed to get cloud", zap.String("cloud", cloud.Name))
-		return errors.E(op, err)
+		return err
 	}
 
-	cr := cloud.Region(modelInfo.CloudRegion)
-	if cr.Name != modelInfo.CloudRegion {
-		return errors.E(op, "cloud region not found")
+	cr := cloud.Region(m.modelInfo.CloudRegion)
+	if cr.Name != m.modelInfo.CloudRegion {
+		return errors.E("cloud region not found")
 	}
 
-	model.CloudRegionID = cr.ID
-	model.CloudRegion = cr
+	m.model.CloudRegionID = cr.ID
+	m.model.CloudRegion = cr
 
-	err = j.Database.AddModel(ctx, &model)
-	if err != nil {
-		if errors.ErrorCode(err) == errors.CodeAlreadyExists {
-			return errors.E(op, err, "model already exists")
+	return nil
+}
+
+func (m *modelImporter) save(ctx context.Context) error {
+	return m.jimm.Database.Transaction(func(d *db.Database) error {
+		err := m.jimm.Database.AddModel(ctx, &m.model)
+		if err != nil {
+			if errors.ErrorCode(err) == errors.CodeAlreadyExists {
+				return fmt.Errorf("model (%s) already exists", m.model.Name)
+			}
+			return err
 		}
+		for _, offer := range m.offersToAdd {
+			var dbOffer dbmodel.ApplicationOffer
+			dbOffer.FromJujuApplicationOfferAdminDetailsV5(offer)
+			dbOffer.ModelID = m.model.ID
+			if err := m.jimm.Database.AddApplicationOffer(ctx, &dbOffer); err != nil {
+				if errors.ErrorCode(err) == errors.CodeAlreadyExists {
+					return fmt.Errorf("offer with URL %s already exists", offer.OfferURL)
+				}
+				return err
+			}
+		}
+		return nil
+	})
+}
+
+// ImportModel imports model with the specified UUID from the controller.
+func (j *JIMM) ImportModel(ctx context.Context, user *openfga.User, controllerName string, modelTag names.ModelTag, newOwner string) error {
+	const op = errors.Op("jimm.ImportModel")
+
+	if err := j.checkJimmAdmin(user); err != nil {
+		return err
+	}
+
+	importer := newModelImporter(j)
+	if err := importer.fetchModelInfo(ctx, controllerName, modelTag); err != nil {
+		return errors.E(op, err)
+	}
+
+	if err := importer.setModelOwner(ctx, newOwner); err != nil {
+		return errors.E(op, err)
+	}
+
+	if err := importer.addPermissions(ctx); err != nil {
+		return errors.E(op, err)
+	}
+
+	// TODO(CSS-5458): Remove the below section on cloud credentials once we no longer persist the relation between
+	// cloud credentials and models.
+	// Update: We need to investigate this further, if a user updates their cloud-credential it will update the credential
```

Comment:
hm. this might be a bit confusing. if a user updates their cloud-credential, sure, jimm will try and update the cloud credential on the controller responsible for this model, but the actual credential used by the model will not be affected. the user would need to use the `juju set-credential` command to set credentials first and then update them to get the actual model cloud credentials updated.

Reply:
Yes this will need fixing I think. For a follow up PR.